Library/OcConsoleLib/GopPassThrough.c

```diff
@@ -0,0 +1,285 @@
+  IN  UINTN                                   DestinationY,
+  IN  UINTN                                   Width,
+  IN  UINTN                                   Height,
+  IN  UINTN                                   Delta         OPTIONAL
+  )
+{
+  OC_GOP_PROTOCOL  *OcGopDraw;
+
+  OcGopDraw = BASE_CR (This, OC_GOP_PROTOCOL, GraphicsOutput);
+
+  return OcGopDraw->Uga->Blt (
+    OcGopDraw->Uga,
+    (EFI_UGA_PIXEL *) BltBuffer,
+    (EFI_UGA_BLT_OPERATION) BltOperation,
+    SourceX,
+    SourceY,
+    DestinationX,
+    DestinationY,
+    Width,
+    Height,
+    Delta
+    );
+}
+
+EFI_STATUS
+OcProvideGopPassThrough (
+  VOID
+  )
+{
+  EFI_STATUS                       Status;
+  UINTN                            HandleCount;
+  EFI_HANDLE                       *HandleBuffer;
+  UINTN                            Index;
+  EFI_GRAPHICS_OUTPUT_PROTOCOL     *GraphicsOutput;
+  EFI_UGA_DRAW_PROTOCOL            *UgaDraw;
+  OC_GOP_PROTOCOL                  *OcGopDraw;
+  APPLE_FRAMEBUFFER_INFO_PROTOCOL  *FramebufferInfo;
+  EFI_PHYSICAL_ADDRESS             FramebufferBase;
+  UINT32                           FramebufferSize;
+  UINT32                           ScreenRowBytes;
+  UINT32                           ScreenWidth;
+  UINT32                           ScreenHeight;
+  UINT32                           ScreenDepth;
+  UINT32                           HorizontalResolution;
+  UINT32                           VerticalResolution;
+  EFI_GRAPHICS_PIXEL_FORMAT        PixelFormat;
+  UINT32                           ColorDepth;
+  UINT32                           RefreshRate;
+
+  Status = gBS->LocateHandleBuffer (
+    ByProtocol,
+    &gEfiGraphicsOutputProtocolGuid,
+    NULL,
+    &HandleCount,
+    &HandleBuffer
+    );
+
+  if (!EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCC: Found %u handles with GOP draw\n", (UINT32) HandleCount));
+    FreePool (HandleBuffer);
+  } else {
+    DEBUG ((DEBUG_INFO, "OCC: Found NO handles with GOP draw\n"));
+  }
+
+  Status = gBS->LocateHandleBuffer (
+    ByProtocol,
+    &gEfiUgaDrawProtocolGuid,
+    NULL,
+    &HandleCount,
+    &HandleBuffer
+    );
+
+  if (!EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCC: Found %u handles with UGA for GOP check\n", (UINT32) HandleCount));
+
+    for (Index = 0; Index < HandleCount; ++Index) {
+      DEBUG ((DEBUG_INFO, "OCC: Trying handle %u - %p\n", (UINT32) Index, HandleBuffer[Index]));
+
+      Status = gBS->HandleProtocol (
+        HandleBuffer[Index],
+        &gEfiUgaDrawProtocolGuid,
+        (VOID **) &UgaDraw
+        );
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_INFO, "OCC: No UGA protocol - %r\n", Status));
+        continue;
+      }
+
+      Status = gBS->HandleProtocol (
+        HandleBuffer[Index],
+        &gEfiGraphicsOutputProtocolGuid,
+        (VOID **) &GraphicsOutput
+        );
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_INFO, "OCC: No GOP protocol - %r\n", Status));
```

Comment:
I suggest dropping this print in favour of a comment in else branch: `Skipping GOP proxying due as GOP is already present on %p` or something similar. The same change could be done to UGA.